numpyro/infer/barker.py

```diff
@@ -0,0 +1,209 @@
+    :param float target_accept_prob: The target acceptance probability that is used to guide
+        step size adapation. Defaults to ``target_accept_prob=0.4``.
+    :param callable init_strategy: a per-site initialization function.
+        See :ref:`init_strategy` section for available functions.
+
+    **Example**
+
+    .. doctest::
+
+        >>> import jax
+        >>> import jax.numpy as jnp
+        >>> import numpyro
+        >>> import numpyro.distributions as dist
+        >>> from numpyro.infer import MCMC, BarkerMH
+
+        >>> def model():
+        ...     numpyro.sample("x", dist.Normal().expand([10]))
+        ...     numpyro.sample("obs", dist.Normal(x, 1.0), obs=jnp.ones(10))
+        >>>
+        >>> kernel = Barker(model)
+        >>> mcmc = MCMC(kernel, num_warmup=1000, num_samples=1000, progress_bar=True)
+        >>> mcmc.run(jax.random.PRNGKey(0))
+        >>> mcmc.print_summary()
+    """
+    def __init__(self, model=None, potential_fn=None, step_size=1.0,
+                 adapt_step_size=True, adapt_mass_matrix=True, dense_mass=False,
+                 target_accept_prob=0.4, init_strategy=init_to_uniform):
+        if not (model is None) ^ (potential_fn is None):
+            raise ValueError('Only one of `model` or `potential_fn` must be specified.')
+        if dense_mass:
+            raise ValueError('Only dense_mass=False is currently supported')
+        self._model = model
+        self._potential_fn = potential_fn
+        self._step_size = step_size
+        self._adapt_step_size = adapt_step_size
+        self._adapt_mass_matrix = adapt_mass_matrix
+        self._dense_mass = dense_mass
+        self._target_accept_prob = target_accept_prob
+        self._init_strategy = init_strategy
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def sample_field(self):
+        return 'z'
+
+    def get_diagnostics_str(self, state):
+        return 'step size {:.2e}. acc. prob={:.2f}'.format(state.adapt_state.step_size,
+                                                           state.mean_accept_prob)
+
+    def _init_state(self, rng_key, model_args, model_kwargs, init_params):
+        if self._model is not None:
+            params_info, potential_fn_gen, self._postprocess_fn, model_trace = initialize_model(
+                rng_key,
+                self._model,
+                dynamic_args=True,
+                init_strategy=self._init_strategy,
+                model_args=model_args,
+                model_kwargs=model_kwargs)
+            init_params = params_info[0]
+            model_kwargs = {} if model_kwargs is None else model_kwargs
+            self._potential_fn = potential_fn_gen(*model_args, **model_kwargs)
+        return init_params
+
+    def init(self, rng_key, num_warmup, init_params, model_args, model_kwargs):
+        self._num_warmup = num_warmup
+        # TODO (low-priority): support chain_method="vectorized", i.e. rng_key is a batch of keys
```

Comment:
do we need to raise errors to catch this?

Reply:
Yeah, that is a good idea! the assertation is `assert rng_key.shape == (2,)`

Reply:
how does this look?